stream_alert/alert_processor/outputs/credentials/provider.py

```diff
@@ -0,0 +1,654 @@
+"""
+Copyright 2017-present, Airbnb Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import json
+import os
+import shutil
+import tempfile
+from abc import abstractmethod
+
+import boto3
+from botocore.exceptions import ClientError
+
+from stream_alert.shared.helpers.boto import default_config
+from stream_alert.shared.logger import get_logger
+from stream_alert_cli.outputs.helpers import kms_encrypt, send_creds_to_s3
+
+LOGGER = get_logger(__name__)
+
+
+class OutputCredentialsProvider(object):
+    """Loads credentials that are housed on AWS S3, or cached locally.
+
+    Helper service to OutputDispatcher.
+
+    OutputDispatcher implementations may require credentials to authenticate with an external
+    gateway. All credentials for OutputDispatchers are to be stored in a single bucket on AWS S3
+    and are encrypted with AWS KMS. When alerts are dispatched via OutputDispatchers, these
+    encrypted credentials are downloaded and cached locally on the filesystem. Then, AWS KMS is
+    used to decrypt the credentials when in use.
+
+    Public methods:
+        load_credentials: Returns a dict of the credentials requested
+        get_local_credentials_temp_dir(): Returns full path to a temporary directory where all
+            encrypted credentials are cached.
+
+    """
+
+    def __init__(self, config, defaults, region, service_name, prefix=None, aws_account_id=None):
+        self._config = config
+        self._region = region
+        self._defaults = defaults
```

Comment:
if changing `defaults` to optional (see other comment), you could do `self._defaults = defaults if defaults else {}` if you want it to always be a dictionary